src/transports/http.rs

```diff
@@ -1,283 +1,173 @@
-
-        let client = client_builder.build()?;
-
-        let basic_auth = {
-            let url = Url::parse(url)?;
-            let user = url.username();
-            let auth = format!("{}:{}", user, url.password().unwrap_or_default());
-            if &auth == ":" {
-                None
-            } else {
-                Some(HeaderValue::from_str(&format!("Basic {}", base64::encode(&auth)))?)
-            }
-        };
-
-        Ok(Http {
-            id: Arc::new(AtomicUsize::new(1)),
-            url: url.parse()?,
-            basic_auth,
-            client,
-        })
+    pub fn new(url: &str) -> Result<Self> {
+        let client = Client::builder()
+            .build()
+            .map_err(|err| Error::Transport(format!("failed to build client: {}", err)))?;
+        Ok(Self::with_client(client, url.parse()?))
     }
 
-    fn send_request<F, O>(&self, id: RequestId, request: rpc::Request, extract: F) -> Response<F>
-    where
-        F: Fn(Vec<u8>) -> O,
-    {
-        let request = helpers::to_string(&request);
-        log::debug!("[{}] Sending: {} to {}", id, request, self.url);
-        let len = request.len();
-
-        let mut request_builder = self
-            .client
-            .post(self.url.clone())
-            .header(
-                reqwest::header::CONTENT_TYPE,
-                HeaderValue::from_static("application/json"),
-            )
-            .header(reqwest::header::USER_AGENT, HeaderValue::from_static("web3.rs"))
-            .body(request);
-
-        // Don't send chunked request
-        if len < MAX_SINGLE_CHUNK {
-            request_builder = request_builder.header(reqwest::header::CONTENT_LENGTH, len.to_string());
+    /// Like `new` but with a user provided client instance.
+    pub fn with_client(client: Client, url: Url) -> Self {
+        Self {
+            client,
+            inner: Arc::new(Inner {
+                url,
+                id: AtomicUsize::new(0),
+            }),
         }
+    }
 
-        // Send basic auth header
-        if let Some(ref basic_auth) = self.basic_auth {
-            request_builder = request_builder.header(reqwest::header::AUTHORIZATION, basic_auth.clone());
-        }
+    fn next_id(&self) -> RequestId {
+        self.inner.id.fetch_add(1, Ordering::AcqRel)
+    }
 
-        let result = request_builder.send();
-        Response::new(id, Box::pin(result), extract)
+    fn new_request(&self) -> (Client, Url) {
+        (self.client.clone(), self.inner.url.clone())
     }
 }
 
+// Id is only used for logging.
+async fn execute_rpc<T: DeserializeOwned>(client: &Client, url: Url, request: &Request, id: RequestId) -> Result<T> {
+    log::debug!("[id:{}] sending request: {:?}", id, serde_json::to_string(&request)?);
+    let response = client
+        .post(url)
+        .json(request)
+        .send()
+        .await
+        .map_err(|err| Error::Transport(format!("failed to send request: {}", err)))?;
+    let status = response.status();
```

Comment:
What happened with:
- basic auth support
- custom user agent
- non-chunked request if they are below `MAX_SINGLE_CHUNK`
- proxy support
- tls support

Is it intended now for the user to use `with_client` method and pre-configure that?

Reply:
From glancing over the reqwest code it already automatically does basic auth and enables proxy support with the same env variable we were using. This is why I removed that part. Although I did not test this.

The ssl function the previous code was using does *not* enable ssl. It forces reqwest to use a specific ssl backend in case several of the ssl features are enabled. I don't see a reason to do this. If a user does enable multiple features it means they probably do not care and if they do care they can still customize the client.

> non-chunked request if they are below MAX_SINGLE_CHUNK

This one I didn't understand the purpose of and the code didn't give a reason. I believe it is better to let the http library decide how to send the request. It knows how long the data is so it can set Content Length as it pleases.

> custom user agent

I thought reqwest already set a user agent but now I see it does not. Will add the rust-web3 user agent back.

Reply:
Thanks for detailed explanations, I'm good with all of them.

The non-chunked request thing was mainly convenience - hyper was defaulting to send a chunked request (without `Content-Length`) and it was a bit more cumbersome to read over a raw TCP stream (chunk decoding). It's fine to let `reqwest` decide, I agree.

> I thought reqwest already set a user agent but now I see it does not. Will add the rust-web3 user agent back.

Cool, thanks!